src/Transports/MassTransit.Azure.ServiceBus.Core/AzureServiceBusTransport/Receiver.cs

```diff
@@ -63,116 +51,129 @@ protected async Task ExceptionHandler(ProcessErrorEventArgs args)
-
-                _deliveryComplete.TrySetResult(false);
+                await _clientContext.NotifyFaulted(args.Exception, args.EntityPath).ConfigureAwait(false);
 
-            #pragma warning disable 4014
-                Task.Run(() => this.Stop($"Receiver Exception: {args.Exception.Message}"));
-            #pragma warning restore 4014
+                TrySetConsumeException(args.Exception);
             }
         }
 
-        Task HandleDeliveryComplete()
-        {
-            if (IsStopping)
-                _deliveryComplete.TrySetResult(true);
-            return Task.CompletedTask;
-        }
-
         protected override async Task StopAgent(StopContext context)
         {
-            await _context.ShutdownAsync().ConfigureAwait(false);
+            await _clientContext.ShutdownAsync().ConfigureAwait(false);
 
-            SetCompleted(ActiveAndActualAgentsCompleted(context)
-                .ContinueWith(_ => Close()));
+            await base.StopAgent(context).ConfigureAwait(false);
 
-            await Completed.ConfigureAwait(false);
-
-            LogContext.Debug?.Log("Receiver stopped: {InputAddress}", _context.InputAddress);
+            LogContext.Debug?.Log("Receiver stopped: {InputAddress}", _clientContext.InputAddress);
         }
 
-        async Task ActiveAndActualAgentsCompleted(StopContext context)
+        protected override async Task ActiveAndActualAgentsCompleted(StopContext context)
         {
-            if (_messageReceiver.ActiveDispatchCount > 0)
-            {
-                try
-                {
-                    await _deliveryComplete.Task.OrCanceled(context.CancellationToken).ConfigureAwait(false);
-                }
-                catch (OperationCanceledException)
-                {
-                    LogContext.Warning?.Log("Stop canceled waiting for message consumers to complete: {InputAddress}", _context.InputAddress);
-                }
-            }
-        }
+            await base.ActiveAndActualAgentsCompleted(context).ConfigureAwait(false);
 
-        async Task Close()
-        {
             try
             {
-                await _context.CloseAsync().ConfigureAwait(false);
+                await _clientContext.CloseAsync().ConfigureAwait(false);
             }
             catch (Exception exception)
             {
-                LogContext.Warning?.Log(exception, "Failed to close the message receiver context: {InputAddress}", _context.InputAddress);
+                LogContext.Warning?.Log(exception, "Failed to close the message receiver context: {InputAddress}", _clientContext.InputAddress);
             }
         }
 
         async Task OnMessage(ProcessMessageEventArgs messageReceiver, ServiceBusReceivedMessage message, CancellationToken cancellationToken)
         {
+            var context = new ServiceBusReceiveContext(message, _context);
+            MessageLockContext lockContext = new ServiceBusMessageLockContext(messageReceiver, message);
```

Comment:
If you move lockContext above context, you can add the payloads (both lockContext and _clientContext) to the constructor of the receive context as object[] params and then it initialized the payload cache in a single operation, eliminate all the GetOrAddPayload overhead.